src/registry.ts

```diff
@@ -168,21 +174,25 @@ function headOk(
 		options.method = "HEAD";
 		request(options, allowInsecure, (res) => {
 			logger.debug(`HEAD ${url}`, res.statusCode);
-			// Not found
-			if (res.statusCode == 404) return resolve(false);
-			// OK
-			if (res.statusCode == 200) return resolve(true);
-			// Redirected
-			if (redirectCodes.includes(res.statusCode ?? 0) && res.headers.location) {
-				if (optimisticCheck) return resolve(true);
-				return resolve(headOk(res.headers.location, headers, allowInsecure, optimisticCheck, ++depth));
-			}
-			// Unauthorized
-			// Possibly related to https://gitlab.com/gitlab-org/gitlab/-/issues/23132
-			if (res.statusCode == 401) {
-				return resolve(false);
-			}
-			reject(toError(res));
+			waitForResponseEnd(res, (data) => {
+				// Not found
+				if (res.statusCode == 404) return resolve(false);
+				// OK
+				if (res.statusCode == 200) return resolve(true);
+				// Redirected
+				if (redirectCodes.includes(res.statusCode ?? 0) && res.headers.location) {
+					if (optimisticCheck) return resolve(true);
+					return resolve(headOk(res.headers.location, headers, allowInsecure, optimisticCheck, ++depth));
+				}
+				// Unauthorized
+				// Possibly related to https://gitlab.com/gitlab-org/gitlab/-/issues/23132
```

Comment:
I can take a look at this. I managed to push to GitLab after getting a jwt token by running 

```
SCOPE="repository:${CI_PROJECT_NAMESPACE}/${CI_PROJECT_NAME}:pull,push"
AUTH_URL="${CI_SERVER_URL}/jwt/auth?service=container_registry&scope=${SCOPE}"
TOKEN=$(curl -sL --user "${CI_REGISTRY_USER}:${CI_REGISTRY_PASSWORD}" "${AUTH_URL}" | sed 's/.*"token":"\{0,1\}\([^,"]*\)"\{0,1\}.*/\1/')
yarn containerify --folder . --toRegistry "https://${CI_REGISTRY}/v2/${CI_PROJECT_NAMESPACE}" --toImage "${CI_PROJECT_NAME}:latest" --toToken $TOKEN
```

I assume we can run some checks on the container registry URL to determine if we should fetch this token or not.